sandbox/apparmor/notify/listener/listener.go

```diff
@@ -424,38 +425,46 @@ func (l *Listener) waitAndRespondAaClassFile(req *Request, msg *notify.MsgNotifi
 	resp := notify.ResponseForRequest(&msg.MsgNotification)
 	resp.MsgNotification.Error = 0 // ignored in responses
 	resp.MsgNotification.NoCache = 1
-	var response *Response
+
+	var explicitlyAllowed uint32 = 0
 	select {
-	case response = <-req.replyChan:
-		break
+	case responsePermission := <-req.replyChan:
+		if responsePermission == nil {
+			// Treat nil as allowing no permission
+			break
+		}
+		perms, ok := responsePermission.(notify.FilePermission)
+		if !ok {
+			// should not occur, Reply() checks that type is correct
+			logger.Debugf("invalid reply from client: %+v; denying request", responsePermission)
+			break
+		}
+		explicitlyAllowed = uint32(perms)
 	case <-l.tomb.Dying():
 		// don't bother sending deny response, kernel will auto-deny if needed
 		return nil
 	}
-	allow := response.Allow
-	perms, ok := response.Permission.(notify.FilePermission)
-	if !ok {
-		// should not occur, Reply() checks that type is correct
-		logger.Debugf("invalid reply from client: %+v; denying request", response)
-		allow = false
-	}
-	if allow {
-		// Allow permissions which AppArmor initially allowed, along with those
-		// which the were initially denied but the user explicitly allowed.
-		// Any permissions which are omitted from both the allow and deny
-		// fields will be automatically denied by the kernel.
-		resp.Allow = msg.Allow | (uint32(perms) & msg.Deny)
-		resp.Deny = 0 // TODO: switch to equivalent but clearer ~uint32(perms) & msg.Deny
-		resp.Error = 0
-	} else {
-		resp.Allow = msg.Allow
-		resp.Deny = msg.Deny
-		resp.Error = msg.Error
-		// msg.Error is field from MsgNotificationResponse, and is unused.
-		// msg.MsgNotification.Error is also ignored in responses.
-	}
+
+	// If the same permission appears in both the allow and deny fields,
+	// treat it as initially denied.
```

Comment:
it may be worth expanding the comment a little bit to mention that this is unexpected but it's also what the kernel gave us

Reply:
I want to wait to chat with JJ before adding that comment, in case it actually is expected and it's only unexpected by *me* :) but yes I agree